Env: mu-plugins and parent directories

noahtallen opened the pull request:
## Description
This PR accomplishes two goals:
1. Add mu-plugins support
2. Allow parent directories as a type of source. 

Resolves #20790
Fixes #20923 (because we no longer use flatMap)

## How has this been tested?
Ran the tool locally in Gutenberg and another plugin I develop.

## Types of changes
New feature.

## Approach
For mu-plugins, specify an array of any valid `Source` in your `.wp-env.json` file:

```json
{
  "mu-plugins": [ "../mu-plugin-thing" ]
}
```

For parent directories, specify a single string instead of an array:
```json
{
  "themes": "../path/to/themes"
}
```

Then, that themes directory is mounted as `wp-content/themes`. This way, you can control the entire plugin/mu-plugin/theme directories if needed.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

noisysocks commented on packages/env/lib/env.js:
```diff
@@ -477,22 +480,26 @@ async function configureWordPress( environment, config ) {
-			`wp plugin activate ${ pluginSource.basename }`,
-			options
-		);
+	if ( Array.isArray( config.pluginSources ) ) {
```
Ah, hm. It's unfortunate that plugins and themes aren't automatically activated if you mount the whole directory. Is there a workaround? Maybe we could activate all plugins using e.g. `wp plugin activate --all`? And the first theme using e.g. `wp theme activate $( wp theme list | head -1 )`? Will need to investigate what the implications are of doing this, though.

noahtallen replied:
That would probably be a cleaner way to support this. The only other way I can think would be to read all the subdirectory basenames and activate them one by one... I feel like we want to avoid that, but I guess I don't have a strong reason to. 

What implications would there be running those commands compared to just activating them one by one?

noahtallen replied:
I have it working essentially how you mentioned. However, I couldn't get `wp theme activate $( wp theme list | head -1 )` working because docker was complaining about too many positional arguments. I did get it working in a less efficient way, but I'll try to put it all in one command :)

noahtallen replied:
Yeah, for some reason piping won't work with docker compose run. I'm not sure why.

Try this command: `wp-env run cli "wp theme list --fields=name | head -2"`. It gives the error: `Too many positional arguments: | head -2`. 

However, this does work: `wp-env run cli wp theme list --fields=name | head -2` (I assume because we're piping locally instead of inside the docker container)

noisysocks replied:
Ah, right. I think it doesn't work because node runs `docker` without bash (which is what handles things like `$()` and `|`) being involved at all. What you did makes sense 👍

noisysocks commented on packages/env/lib/config.js:
```diff
@@ -143,23 +145,8 @@ module.exports = {
-			);
-		}
+		// Throw an error if any of the sources are invalid.
+		validateSources( config, [ 'plugins', 'themes', 'mu-plugins' ] );
```
I get that without this abstraction there's a lot of repetition but, to me, this feels a little _too_ abstract. Maybe a compromise would be to remove the `forEach` and call `validateSources` once per source?

```js
validateSources( config.plugins );
validateSources( config.themes );
validateSources( config[ 'mu-plugins' ] );
```

noahtallen replied:
Ok :) This now looks like `validateSources( config.plugins, 'plugins' );` since we need to know "name" of the source type so we can inform the user which property needs to be fixed. An alternative is `validateSources( config, 'plugins' );` (where we would read the value from the config inside the function). What do you think?

noisysocks replied:
I prefer how you have it 🙂

noisysocks commented on packages/env/lib/config.js:
```diff
@@ -206,21 +193,60 @@ module.exports = {
+				config[ 'mu-plugins' ],
+				workDirectoryPath
 			),
+			themeSources: getSources( config.themes, workDirectoryPath ),
```
I'm not huge on the extra little layer of indirection that `getSources` adds. I'd be tempted to pull in lodash so that we can use [`castArray`](https://lodash.com/docs/4.17.15#castArray). I appreciate that this is straying into individual preference territory, though 😀 What do you think?

```js
pluginSources: castArray( config.plugins ).map( ( sourceString ) =>
	parseSourceString( sourceString, {
		workDirectoryPath,
	} )
),
themeSources: castArray( config.themes ).map( ( sourceString ) =>
	parseSourceString( sourceString, {
		workDirectoryPath,
	} )
),
muPluginSources: castArray( config[ 'mu-plugins' ] ).map( ( sourceString ) =>
	parseSourceString( sourceString, {
		workDirectoryPath,
	} )
),
```

noahtallen replied:
This won't work with the current implementation because we let `muPluginSources` be `string` or `string[]`. In the `string` case, the corresponding directory is mounted as `wp-content/themes` (for example). In the array case, all of the directories in the array are mounted as `wp-content/themes/dir`. So we need to be able to switch between those two cases

noisysocks replied:
Right you are!

noisysocks replied:
Right you are!

noisysocks commented on packages/env/lib/env.js:
```diff
@@ -70,6 +71,20 @@ module.exports = {
 					.join( '\n' );
 		};
 
+		const downloader = ( source ) =>
```
Function names should start with a verb. `getSourceDownloder`?

noahtallen replied:
done!

noisysocks commented on packages/env/lib/env.js:
```diff
@@ -91,21 +106,9 @@ module.exports = {
-			),
+			...getSourceDownloads( config.pluginSources, downloader ),
+			...getSourceDownloads( config.muPluginsSources, downloader ),
+			...getSourceDownloads( config.themeSources, downloader ),
```
Not sure if it's because I'm dumb (I am!) or because it's Monday morning (it is!) but I struggled to parse this at first. Similar to my other comment above, I'd be tempted to use `castArray` here to remove `getSourceDownloads`.

```js
...castArray( config.pluginSources ).map( downloader ),
...castArray( config.muPluginsSources ).map( downloader ),
...castArray( config.themeSources ).map( downloader ),
```

noahtallen replied:
cast array would work very nicely here! I didn't know about it before :)

noahtallen replied:
and done

noisysocks reviewed (changes requested):
Browsed through the code and this is looking good! I haven't tested it yet. I think we need to sort out what to do with plugin and theme activation—it would be pretty inconvenient to have to activate things manually every time you start the server.

noahtallen commented on packages/env/README.md:
```diff
@@ -210,15 +210,16 @@ You can customize the WordPress installation, plugins and themes that the develo
-| `"port"`      | `integer`      | `8888`                                   | The primary port number to use for the insallation. You'll access the instance through the port: 'http://localhost:8888'. |
-| `"testsPort"` | `integer`      | `8889`                                   | The port number to use for the tests instance.                                                                            |
-| `"config"`    | `Object`      | `"{ WP_DEBUG: true, SCRIPT_DEBUG: true }"` | Mapping of wp-config.php constants to their desired values.                                                               |
+| `"plugins"`   | `string[]|string`    | `[]`                                | A list of plugins to install and activate in the environment. To mount a directory as `wp-content/plugins`, the value should be a path to the applicable directory instead of an array. |
```
Done

noisysocks commented on packages/env/lib/env.js:
```diff
@@ -477,20 +474,30 @@ async function configureWordPress( environment, config ) {
 		await dockerCompose.run(
 			environment === 'development' ? 'cli' : 'tests-cli',
-			`wp theme activate ${ themeSource.basename }`,
+			`wp theme activate ${ firstTheme }`,
```
Let's use `dockerCompose.run()`'s array argument form—this makes sure things work properly if `firstTheme` contains spaces for some reason.

```suggestion
			[ 'wp', 'theme', 'activate', firstTheme ],
```

noahtallen replied:
done

noisysocks commented on packages/env/lib/env.js:
```diff
@@ -477,20 +474,30 @@ async function configureWordPress( environment, config ) {
-	const [ themeSource ] = config.themeSources;
-	if ( themeSource ) {
+	// Index 0 contains "name" (the header of the output)
+	const firstTheme = themeOutput.out.split( '\n' )[ 1 ];
```
Could use destructuring here if you wanted to remove that magic number—I hear it's all the rage 😎

```suggestion
	const [ , firstTheme ] = themeOutput.out.split( '\n' );
```

noahtallen replied:
😎 I'll drink that javascript juice 🍹

noisysocks commented on packages/env/lib/build-docker-compose-config.js:
```diff
@@ -1,13 +1,28 @@
+ * @typedef {import('./config').Source} Source
  */
-const fs = require( 'fs' );
-const path = require( 'path' );
```
Maybe a bad rebase? Removing these imports is causing some errors.

noahtallen replied:
yep! when I first started this PR, we were able to remove these deps, but in the meantime we added some other functionality which uses them :D

noisysocks reviewed (changes requested):
Looking really good! Just need to fix up the `✖ fs is not defined` error I'm getting when I run `wp-env start`.

noahtallen commented:
Alright, this should be ready for another look. I also added two more improvements:
- cleaned up the way we check for the environment in the WP install step
- added mu-plugins to the new tests mount format

noisysocks reviewed (approved):
Looks good! Nice work 👍

noahtallen commented:
@noisysocks not sure why, but the build in Travis fails running `npm i` and `npm run check-local-changes`. The branch is rebased and those commands complete successfully locally, so I'm not sure what more there is to try 🤷‍♂

noisysocks commented:
It might be that you ran `npm install` with a different version of `npm` or `node` than what Travis CI is using. I ran `nvm install --latest-npm; npm install` locally and committed the generated `package-lock.json`. That seems to have made Travis CI happy.

noisysocks commented:
Hm, I just realised that the approach of activating all plugins using `wp plugin activate --all` will mean that all of the Gutenberg test plugins (`packages/e2e-tests/plugins/`) will be activated by default. That's probably not what we want for the default Gutenberg development environment.

Is there an alternative way of activating just the plugins that are specified in `"plugins"`? Maybe, if `"plugins"` or `"themes"` is a single path we could "expand" it into an array of paths by searching the specified directory for things that look like plugins or themes.

noisysocks commented on packages/env/lib/env.js:
```diff
@@ -469,28 +468,34 @@ async function configureWordPress( environment, config ) {
-	const [ themeSource ] = config.themeSources;
-	if ( themeSource ) {
+	// Index 0 contains "name" (the header of the output), so use next index.
+	const [ , firstTheme ] = themeOutput.out.split( '\n' );
```
Are we guaranteed that the first theme listed by `wp theme list` is the first theme specified in `.wp-env.json`'s `"themes"` array? It may be that WP-CLI lists themes alphabetically.

noahtallen replied:
good point. I changed it to still activate the first theme in the array if we are using the array format.

I wonder, do we want this functionality? If the user activates a different theme while using wp-env, wouldn't this overwrite their choice every time they restart the container?

noisysocks replied:
That's a good question. I think it makes sense to have `wp-env` activate a theme because it mirrors what it does for plugins. But now you have me questioning myself... 😀

noahtallen replied:
Yeah.. I mean, either are easy enough to activate internally. I guess the difference is that if you list all plugins, you probably want them all to be running most of the time. But with themes, only one can run at a time, so why not let the user choose in wp-admin? (I bet we'll get a "bug report" about the behavior at some point ;))

noisysocks commented on packages/env/lib/env.js:
```diff
@@ -469,28 +468,34 @@ async function configureWordPress( environment, config ) {
-		);
+	await dockerCompose.run(
+		containerName,
+		'wp plugin activate --all',
```
I just realised that this approach of activating all plugins using `wp plugin activate --all` will mean that all of the Gutenberg test plugins (`packages/e2e-tests/plugins`) will be activated by default. That's probably not what we want for the default Gutenberg development environment.

I'm not sure of the best way around this. Some ideas:

1. We could look at adding support for plugins that are not activated automatically.
2. We could look at adding support for plugins that are only loaded into the tests environment. 
3. We could go back to hardcoding this for Gutenberg.

noahtallen replied:
That's an interesting conundrum. My intuition would be to do number 2:

- number 3 isn't an ideal choice any way you cut it
- number 1 means we wouldn't directly support test environment mu-plugins. I assume we want to automatically activate them for the test environment :)

So for the api, maybe we change the source format in wp-env to accept `string | object` where the string is a normal source string (current approach) and the object has the format:

```js
{
  "source": SourceString,
  "activateInEnv": "development" | "tests" | "none"
}
```

Then, we activate the plugin depending on the environment, or don't activate it at all if none is passed. (just an idea on how to make this a bit more flexible in the long run).

BTW, will this approach work out of the box with the current docker setup? i.e. does `test-cli wp activate plugin` only activate the plugin on the test WP and not on the production one?

noisysocks replied:
> * I assume we want to automatically activate them for the test environment :)

The E2E tests themselves are responsible for activating the plugins that they need, e.g:

https://github.com/WordPress/gutenberg/blob/ea62337a254c52305ee6bb46e00656d665123cc3/packages/e2e-tests/specs/editor/plugins/container-blocks.test.js#L15-L17

> So for the api, maybe we change the source format in wp-env to accept string | object where the string is a normal source string (current approach) and the object has the format:

I'm not a huge fan on the extra weight this adds to `.wp-env.json`. I worry about making this file so complex that it doesn't provide much benefit over using `docker-compose.yml`.

> BTW, will this approach work out of the box with the current docker setup? i.e. does test-cli wp activate plugin only activate the plugin on the test WP and not on the production one?

Yes, that will work because the instances use different databases.

noisysocks replied:
I'm leaning towards option one because it maintains one principal of `wp-env` that I like which is that the two instances it spins up are identical: same core code, same plugins, same themes. It's really nice being able to make this assumption when e.g. writing and debugging E2E tests.

In other words, I think let's add some way of defining a plugin that isn't activated, similar to what you suggested above:

```json
{
	"plugins": [
		".",
		{ "source": "./packages/e2e-tests/plugins", "active": false }
	]
}
```

This is tough to implement, though. We can't use `wp plugin activate --all` since that will activate everything. We also can't iterate through the plugins with `.active = true` and perform `wp plugin activate ${plugin}` because some of the paths will be to _directories containing plugins_ instead of to plugins themselves. We might have to do some weird stuff involving `wp plugin list`, `wp plugin path`, and `wp plugin activate`.

😕

noahtallen replied:
I like that idea. For example, active could default to true. I'm exploring this in another branch -- I think it's a good opportunity to make sure directory support works in general too.

noahtallen replied:
what about mu-plugins that only need to be activated in the test environment?

noahtallen replied:
https://github.com/WordPress/gutenberg/pull/21710

noisysocks reviewed (changes requested):
Sorry for my last minute change of mind here—there's two potential problems with this approach that we ought to consider 😅

noahtallen committed:
Add mu-plugins and top-level directory support

- Sources can be a single source or array
  of sources. Single sources map to the
  matching wp-content directory. Arrays
  of sources are each placed inside.
- "mu-plugins" can be used in the config
  object with the same source string format
  as themes and plugins.

noahtallen committed:
Handle directories when activating themes/plugins

- Use `wp plugin activate --all` for plugins
- Use first listed theme from wp theme list

noahtallen committed:
Specify Gutenberg e2e plugins with new config options

noahtallen committed:
Add tests for new config options

noahtallen committed:
Update documentation

noahtallen committed:
Await get theme

noahtallen commented:
@noisysocks I think this is ready for another look (just rebased on master and verified some things)

see https://github.com/WordPress/gutenberg/pull/21710 for my investigation into adding support for an `active` property.

The only outstanding questions I see are:
1. Should we be activating the same theme every time we run wp-env start? What if users change it? Currently, we constantly override their choice.
2. How do we handle mu-plugins which should only be loaded in one environment, like the Gutenberg e2e test mu plugins?

I'm not sure if either of those questions should block the PR though ;)

epiqueras commented:
## Environment Differences

I think it's clear now that the **different requirements of the development and tests environments require an API like that of [`babelrc`](https://babeljs.io/docs/en/6.26.3/babelrc#env-option)** where you can specify environment-specific options to be merged into the top-level options. We would support `env.development` and `env.tests`, but in the future, we could even support custom environments that would spin up additional WP instances if that's a feature people would find useful.

In Gutenberg, for example, **we can't load `mu-plugins` in the development environment**. They do things like normalizing theme styles, and that would break the expectations of theme developers.

## Theme Activation

Choosing a theme to **activate for the user based on a list's order is not a nice explicit API**. Also, overwriting their **current theme every time they run the command could get annoying**.

I think that we shouldn't activate any theme by default but instead **have a new top-level key, `theme`, or `activeTheme` that allows people to opt-in** to activating a specific theme.

In Gutenberg, for example, we wouldn't want the development theme to be anything specific, and we wouldn't want to overwrite it every time we ran `wp-env`. This is easy to achieve with the proposed `env` option. We would not use `env.development.theme`, but we would keep a consistent theme for tests using `env.tests.theme`. 

*Adding an extra key to the theme sources is also an option, but it's more verbose and opens up questions of what happens when two or more theme sources have the same environment key.*

noahtallen commented:
That's a fair analysis, thanks @epiqueras! I tend to agree to with you about being able to specify an environment for the different options. I'm pondering the situation, and wonder if there is any concern about making the API too complex? This does seem like a very basic requirement for most tools like this, though. I know @noisysocks has mentioned that he would like the dev and tests environment to be as similar in possible, so I'm curious to hear his thoughts about it.

> but instead have a new top-level key, theme, or activeTheme that allows people to opt-in to activating a specific theme

I think this would definitely be a much better API than the implicit one we have now.

noahtallen commented:
@epiqueras in the mean time, I still think we want to add parent directory/mu-plugin support. (Additionally, one could add an override file to not load any of the test mu plugins as a workaround for now.) I'll rebase this and add some of the suggestions (like not loading themes). Then we could review/ship this and (IMO) abandon #21710

noahtallen commented:
Got some feedback here: https://github.com/WordPress/gutenberg/issues/21595#issuecomment-622562088

Where they mentioned that their exceptions for a wp-content directory would not  _just_ be to load the content directory, but also to load any other specified plugins and items. Currently, this PR wouldn't work with that approach, since we treat the `plugins: "path/to/plugins"` case as a content directory implicitly. So we would need to add a way to explicitly add a "content directory" via config, which would probably mean extending the type accepted by `plugins/themes/mu-plugins` from strings to also accept objects as in #21710.

epiqueras commented:
I think that instead of:

>Allow parent directories as a type of source.

We should add `mappings` and let people pass custom directory mappings:

```json
{
  "mappings": { "wp-content/themes": "../path/to/themes" }
}
```

This will be way more flexible, easier to understand, it can be done in parallel to what we already do with `themes/plugins/mu-plugins`, and we avoid creating a complicated schema for the config.

noahtallen commented:
Neat idea :) the only reasons I could see not going this route:

- it’s basically just exposing docker volumes; should we just call it that?
- what value do the themes/plugins/mu-plugins hold now other than back compat?
- is it too advanced and could it allow for more complex environments than we want to support?

epiqueras commented:
>it’s basically just exposing docker volumes; should we just call it that?

I think "volumes" is a less understood term, but the transparency might be worth it. @noisysocks, what do you think?

>what value do the themes/plugins/mu-plugins hold now other than back compat?

They let you combine multiple individual themes/plugins from a multitude of different sources and do more out of the box like activation. Mappings are a power user feature.

>is it too advanced and could it allow for more complex environments than we want to support?

It's better than supporting a bloated and hard to document configuration schema. I don't think its too advanced if we dumb it down to "mapping folders" and avoid getting into more complex docker volume options.

noisysocks commented:
`mappings` sounds like it could be a good escape for advanced use cases.

What does Docker do, though, if you try to map a volume which has an ancestor that is already mapped?

```json
{
	"plugins": [ "/foo" ],
	"mappings": {
		"wp-content/plugins": "/bar"
	}
}
```

Will the plugin in `/foo` be loaded?

epiqueras commented:
Yeah, I think we just need to order them correctly.

noahtallen commented:
cool. if i get a chance, I'll refactor my PRs and stuff to work with this approach. I hope I can do that before next week?

noahtallen commented:
Closing in favor of #22256 (for parent directories) and TBD (for mu-plugins)

> Will the plugin in /foo be loaded?

I tested this in #22256 and it does work. We also currently map `/var/www/html` as wordpress' home directory, and everything else we map currently goes into subdirectories of that volume, so that's an example of it working :)
